src/Telegram.Bot/TelegramBotClient.cs

```diff
@@ -104,54 +113,130 @@ public TelegramBotClient(string token, HttpClient? httpClient = null)
+                            stringifiedResponse
+                        );
+                    }
+
+                    throw ExceptionParser.Parse(
+                        failedApiResponse.ErrorCode,
+                        failedApiResponse.Description,
+                        failedApiResponse.Parameters
+                    );
+                }
+
+                var successfulApiResponse = await httpResponse
+                    .DeserializeContentAsync<SuccessfulApiResponse<TResult>>(
+                        actualResponseStatusCode
+                    );
+
+                if (successfulApiResponse is null)
+                {
+                    var stringifiedResponse = await httpResponse.Content.ReadAsStringAsync()
+                        .ConfigureAwait(false);
+
+                    throw new RequestException(
+                        "Required properties not found in response.",
+                        actualResponseStatusCode,
+                        stringifiedResponse
+                    );
+                }
 
-                throw new ApiRequestException("Request timed out", 408, e);
+                return successfulApiResponse.Result;
             }
+            finally
+            {
+                httpResponse?.Dispose();
+            }
+        }
 
-            // required since user might be able to set new status code using following event arg
-            var actualResponseStatusCode = httpResponse.StatusCode;
-            string responseJson = await httpResponse.Content.ReadAsStringAsync()
-                .ConfigureAwait(false);
+        /// <inheritdoc />
+        public async Task<ApiResponse<TResult>> SendRequestAsync<TResult>(
+            IRequest<TResult> request,
+            CancellationToken cancellationToken = default)
+        {
+            var url = _baseRequestUrl + request.MethodName;
 
-            ApiResponseReceived?.Invoke(this, new ApiResponseEventArgs(httpResponse, reqDataArgs));
-
-            switch (actualResponseStatusCode)
-            {
-                case HttpStatusCode.OK:
-                    break;
-                case HttpStatusCode.Unauthorized:
-                case HttpStatusCode.BadRequest when !string.IsNullOrWhiteSpace(responseJson):
-                case HttpStatusCode.Forbidden when !string.IsNullOrWhiteSpace(responseJson):
-                case HttpStatusCode.Conflict when !string.IsNullOrWhiteSpace(responseJson):
-                    // Do NOT throw here, an ApiRequestException will be thrown next
-                    break;
-                default:
-                    httpResponse.EnsureSuccessStatusCode();
-                    break;
+            var httpRequest = new HttpRequestMessage(request.Method, url)
+            {
+                Content = request.ToHttpContent()
+            };
+
+            var reqDataArgs = new ApiRequestEventArgs(request.MethodName, httpRequest.Content);
+            MakingApiRequest?.Invoke(this, reqDataArgs);
```

Comment:
```suggestion
            MakingApiRequest?.Invoke(this, new ApiRequestEventArgs(request.MethodName, httpRequest.Content));
```

Avoid allocating EventArgs if there are no subscribers (should be the same in `MakeRequestAsync` above

Reply:
I've look into it and there might be a downside with that particular solution. With the current setup one could actually track requests because the same instance of `reqDataArgs` then passed inside `ApiResponseEventArgs`. So one could just store event args from `MakingApiRequest` in a dictionary (e.g. with additional metadata like req start time) and on `ApiResponseReceived` associated metadata because of reference equality. I don't know if that was Poulad's original idea, but that's how it's ended up.

Reply:
That's right @tuscen. The intent was to have a way for users to correlate these 2 request and response events. Object reference is kinda a correlation ID between these 2 separate events.